File: src/lexwatt/jcs.py
"""RFC 8785 (JCS) canonical JSON with the LexWatt integer-only numeric profile.

Wire rules (spec §4.1):
- UTF-8, no BOM, no duplicate keys, no lone surrogates, no trailing data.
- Object keys are emitted sorted by UTF-16 code units.
- JSON numbers are restricted to canonical integers: ``-?(0|[1-9][0-9]*)``
  with absolute value <= 2147483647.  No exponent, fraction, leading zero,
  or negative zero.  Larger quantities travel as ``U`` decimal strings.
- Frame depth is bounded (32) before acting on the document.
"""

from __future__ import annotations

from .errors import LexwattError

MAX_DEPTH = 32
MAX_JSON_INT = 2147483647


def _utf16_sort_key(s: str):
    return s.encode("utf-16-be", "surrogatepass")


def _escape_string(s: str, out: list[str]) -> None:
    out.append('"')
    for ch in s:
        o = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\b":
            out.append("\\b")
        elif ch == "\f":
            out.append("\\f")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif o < 0x20:
            out.append("\\u%04x" % o)
        elif 0xD800 <= o <= 0xDFFF:
            raise LexwattError("INVALID_INPUT")
        else:
            out.append(ch)
    out.append('"')


def _write(value, out: list[str]) -> None:
    if value is None:
        out.append("null")
    elif value is True:
        out.append("true")
    elif value is False:
        out.append("false")
    elif isinstance(value, int):
        out.append(str(value))
    elif isinstance(value, str):
        _escape_string(value, out)
    elif isinstance(value, list):
        out.append("[")
        for i, item in enumerate(value):
            if i:
                out.append(",")
            _write(item, out)
        out.append("]")
    elif isinstance(value, dict):
        out.append("{")
        keys = sorted(value.keys(), key=_utf16_sort_key)
        for i, k in enumerate(keys):
            if i:
                out.append(",")
            if not isinstance(k, str):
                raise LexwattError("INVALID_INPUT")
            _escape_string(k, out)
            out.append(":")
            _write(value[k], out)
        out.append("}")
    else:
        raise LexwattError("INVALID_INPUT")


def dumps(value) -> bytes:
    """Serialize to canonical UTF-8 bytes."""
    out: list[str] = []
    _write(value, out)
    return "".join(out).encode("utf-8")


def dumps_str(value) -> str:
    return dumps(value).decode("utf-8")


_HEX = "0123456789abcdef"


class _Parser:
    def __init__(self, data: bytes):
        self.data = data
        self.pos = 0
        self.n = len(data)

    def fail(self) -> LexwattError:
        return LexwattError("INVALID_INPUT")

    def peek(self) -> int:
        return self.data[self.pos] if self.pos < self.n else -1

    def take(self) -> int:
        b = self.data[self.pos]
        self.pos += 1
        return b

    def ws(self) -> None:
        while self.pos < self.n and self.data[self.pos] in (0x20, 0x09, 0x0A, 0x0D):
            self.pos += 1

    def parse(self):
        self.ws()
        v = self.value(0)
        self.ws()
        if self.pos != self.n:
            raise self.fail()
        return v

    def value(self, depth: int):
        if depth > MAX_DEPTH:
            raise self.fail()
        b = self.peek()
        if b < 0:
            raise self.fail()
        if b == 0x7B:  # {
            return self.obj(depth)
        if b == 0x5B:  # [
            return self.arr(depth)
        if b == 0x22:  # "
            return self.string()
        if b == 0x74:  # t
            self.expect(b"true")
            return True
        if b == 0x66:
            self.expect(b"false")
            return False
        if b == 0x6E:
            self.expect(b"null")
            return None
        if b == 0x2D or 0x30 <= b <= 0x39:
            return self.number()
        raise self.fail()

    def expect(self, lit: bytes) -> None:
        if self.data[self.pos : self.pos + len(lit)] != lit:
            raise self.fail()
        self.pos += len(lit)

    def number(self) -> int:
        start = self.pos
        if self.peek() == 0x2D:
            self.pos += 1
        if self.peek() == 0x30:
            self.pos += 1
            # canonical integer grammar forbids any digit after a leading zero
            if 0x30 <= self.peek() <= 0x39:
                raise self.fail()
        elif 0x31 <= self.peek() <= 0x39:
            while 0x30 <= self.peek() <= 0x39:
                self.pos += 1
        else:
            raise self.fail()
        if self.peek() in (0x2E, 0x65, 0x45):  # . e E
            raise self.fail()
        text = self.data[start : self.pos].decode("ascii")
        v = int(text)
        if abs(v) > MAX_JSON_INT:
            raise self.fail()
        return v

    def string(self) -> str:
        assert self.take() == 0x22
        out: list[str] = []
        while True:
            if self.pos >= self.n:
                raise self.fail()
            b = self.take()
            if b == 0x22:
                break
            if b == 0x5C:  # backslash
                if self.pos >= self.n:
                    raise self.fail()
                e = self.take()
                if e == 0x22:
                    out.append('"')
                elif e == 0x5C:
                    out.append("\\")
                elif e == 0x2F:
                    out.append("/")
                elif e == 0x62:
                    out.append("\b")
                elif e == 0x66:
                    out.append("\f")
                elif e == 0x6E:
                    out.append("\n")
                elif e == 0x72:
                    out.append("\r")
                elif e == 0x74:
                    out.append("\t")
                elif e == 0x75:
                    out.append(self.u_escape())
                else:
                    raise self.fail()
            elif b < 0x20:
                raise self.fail()
            elif b < 0x80:
                out.append(chr(b))
            else:
                out.append(self.utf8_char(b))
        s = "".join(out)
        self.check_scalars(s)
        return s

    def u_escape(self) -> str:
        if self.pos + 4 > self.n:
            raise self.fail()
        hexs = self.data[self.pos : self.pos + 4]
        self.pos += 4
        try:
            cp = int(hexs.decode("ascii"), 16)
        except ValueError:
            raise self.fail()
        if 0xD800 <= cp <= 0xDBFF:
            # high surrogate must be followed by \uXXXX low surrogate
            if self.data[self.pos : self.pos + 2] != b"\\u":
                raise self.fail()
            self.pos += 2
            if self.pos + 4 > self.n:
                raise self.fail()
            hexs2 = self.data[self.pos : self.pos + 4]
            self.pos += 4
            try:
                cp2 = int(hexs2.decode("ascii"), 16)
            except ValueError:
                raise self.fail()
            if not (0xDC00 <= cp2 <= 0xDFFF):
                raise self.fail()
            cp = 0x10000 + ((cp - 0xD800) << 10) + (cp2 - 0xDC00)
        return chr(cp)

    def utf8_char(self, first: int) -> str:
        # input was already validated as UTF-8 at the byte level by loads();
        # decode a multi-byte sequence here
        need = 0
        if 0xC0 <= first <= 0xDF:
            need = 1
        elif 0xE0 <= first <= 0xEF:
            need = 2
        elif 0xF0 <= first <= 0xF7:
            need = 3
        else:
            raise self.fail()
        seq = bytes([first]) + self.data[self.pos : self.pos + need]
        self.pos += need
        try:
            return seq.decode("utf-8")
        except UnicodeDecodeError:
            raise self.fail()

    @staticmethod
    def check_scalars(s: str) -> None:
        for ch in s:
            o = ord(ch)
            if 0xD800 <= o <= 0xDFFF:
                raise LexwattError("INVALID_INPUT")

    def obj(self, depth: int):
        self.pos += 1  # {
        out = {}
        self.ws()
        if self.peek() == 0x7D:
            self.pos += 1
            return out
        while True:
            self.ws()
            if self.peek() != 0x22:
                raise self.fail()
            k = self.string()
            self.ws()
            if self.take() != 0x3A:  # :
                raise self.fail()
            self.ws()
            v = self.value(depth + 1)
            if k in out:
                raise LexwattError("INVALID_INPUT")  # duplicate key
            out[k] = v
            self.ws()
            b = self.take()
            if b == 0x7D:
                return out
            if b != 0x2C:
                raise self.fail()

    def arr(self, depth: int):
        self.pos += 1  # [
        out = []
        self.ws()
        if self.peek() == 0x5D:
            self.pos += 1
            return out
        while True:
            self.ws()
            out.append(self.value(depth + 1))
            self.ws()
            b = self.take()
            if b == 0x5D:
                return out
            if b != 0x2C:
                raise self.fail()


def loads(data: bytes):
    """Strict decode of canonical JSON bytes under the LexWatt profile."""
    if not isinstance(data, (bytes, bytearray)):
        raise LexwattError("INVALID_INPUT")
    raw = bytes(data)
    if raw.startswith(b"\xef\xbb\xbf"):
        raise LexwattError("INVALID_INPUT")
    try:
        raw.decode("utf-8")  # rejects invalid UTF-8 and lone surrogates
    except UnicodeDecodeError:
        raise LexwattError("INVALID_INPUT")
    return _Parser(raw).parse()


def loads_str(text: str):
    if not isinstance(text, str):
        raise LexwattError("INVALID_INPUT")
    try:
        data = text.encode("utf-8")
    except UnicodeEncodeError:
        raise LexwattError("INVALID_INPUT")
    return loads(data)

File: ts/src/jcs.ts
/** RFC 8785 (JCS) canonical JSON under the LexWatt integer-only numeric
 * profile (spec §4.1).
 *
 * JS string comparison is already UTF-16 code-unit order, so `sort()`
 * without a comparator is exactly the JCS key ordering.  Numbers are
 * canonical integers with |v| <= 2147483647; booleans are legal JSON.
 */

import { LexwattError } from "./errors.ts";

export const MAX_DEPTH = 32;
export const MAX_JSON_INT = 2147483647;

type Json = null | boolean | number | string | Json[] | { [k: string]: Json };

export function dumps(v: Json): string {
  const parts: string[] = [];
  write(v, parts);
  return parts.join("");
}

export function dumpsBytes(v: Json): Uint8Array {
  return new TextEncoder().encode(dumps(v));
}

function escape(s: string, out: string[]): void {
  out.push('"');
  for (const ch of s) {
    const o = ch.codePointAt(0)!;
    if (ch === '"') out.push('\\"');
    else if (ch === "\\") out.push("\\\\");
    else if (ch === "\b") out.push("\\b");
    else if (ch === "\f") out.push("\\f");
    else if (ch === "\n") out.push("\\n");
    else if (ch === "\r") out.push("\\r");
    else if (ch === "\t") out.push("\\t");
    else if (o < 0x20) out.push("\\u" + o.toString(16).padStart(4, "0"));
    else out.push(ch);
  }
  out.push('"');
}

function write(v: Json, out: string[]): void {
  if (v === null) {
    out.push("null");
  } else if (v === true) {
    out.push("true");
  } else if (v === false) {
    out.push("false");
  } else if (typeof v === "number") {
    if (!Number.isInteger(v) || Math.abs(v) > MAX_JSON_INT) {
      throw new LexwattError("INVALID_INPUT");
    }
    out.push(String(v));
  } else if (typeof v === "string") {
    escape(v, out);
  } else if (Array.isArray(v)) {
    out.push("[");
    v.forEach((item, i) => {
      if (i) out.push(",");
      write(item, out);
    });
    out.push("]");
  } else if (typeof v === "object") {
    out.push("{");
    // JS string < compares UTF-16 code units — exactly JCS ordering
    const keys = Object.keys(v).sort();
    keys.forEach((k, i) => {
      if (i) out.push(",");
      escape(k, out);
      out.push(":");
      write(v[k]!, out);
    });
    out.push("}");
  } else {
    throw new LexwattError("INVALID_INPUT");
  }
}

class Parser {
  pos = 0;
  readonly s: string;
  constructor(s: string) {
    this.s = s;
  }

  fail(): never {
    throw new LexwattError("INVALID_INPUT");
  }

  ws(): void {
    while (this.pos < this.s.length && " \t\n\r".includes(this.s[this.pos]!)) this.pos++;
  }

  parse(): Json {
    this.ws();
    const v = this.value(0);
    this.ws();
    if (this.pos !== this.s.length) this.fail();
    return v;
  }

  value(depth: number): Json {
    if (depth > MAX_DEPTH) this.fail();
    const c = this.s[this.pos];
    if (c === undefined) this.fail();
    if (c === "{") return this.obj(depth);
    if (c === "[") return this.arr(depth);
    if (c === '"') return this.string();
    if (c === "t") return this.lit("true", true);
    if (c === "f") return this.lit("false", false);
    if (c === "n") return this.lit("null", null);
    if (c === "-" || (c >= "0" && c <= "9")) return this.number();
    this.fail();
  }

  lit(word: string, val: Json): Json {
    if (!this.s.startsWith(word, this.pos)) this.fail();
    this.pos += word.length;
    return val;
  }

  number(): number {
    const start = this.pos;
    if (this.s[this.pos] === "-") this.pos++;
    if (this.s[this.pos] === "0") {
      this.pos++;
      const nx = this.s[this.pos];
      if (nx !== undefined && nx >= "0" && nx <= "9") this.fail();
    } else {
      const c = this.s[this.pos];
      if (c === undefined || c < "1" || c > "9") this.fail();
      while (this.s[this.pos] !== undefined && this.s[this.pos]! >= "0" && this.s[this.pos]! <= "9")
        this.pos++;
    }
    const c = this.s[this.pos];
    if (c === "." || c === "e" || c === "E") this.fail();
    const v = Number(this.s.slice(start, this.pos));
    if (!Number.isInteger(v) || Math.abs(v) > MAX_JSON_INT) this.fail();
    return v;
  }

  string(): string {
    if (this.s[this.pos] !== '"') this.fail();
    this.pos++;
    const out: string[] = [];
    for (;;) {
      if (this.pos >= this.s.length) this.fail();
      const c = this.s[this.pos]!;
      this.pos++;
      if (c === '"') break;
      if (c === "\\") {
        const e = this.s[this.pos];
        this.pos++;
        if (e === '"') out.push('"');
        else if (e === "\\") out.push("\\");
        else if (e === "/") out.push("/");
        else if (e === "b") out.push("\b");
        else if (e === "f") out.push("\f");
        else if (e === "n") out.push("\n");
        else if (e === "r") out.push("\r");
        else if (e === "t") out.push("\t");
        else if (e === "u") out.push(this.uEscape());
        else this.fail();
      } else if (c < " ") {
        this.fail();
      } else {
        out.push(c);
      }
    }
    return out.join("");
  }

  uEscape(): string {
    const hex = this.s.slice(this.pos, this.pos + 4);
    if (!/^[0-9a-fA-F]{4}$/.test(hex)) this.fail();
    this.pos += 4;
    let cp = parseInt(hex, 16);
    if (cp >= 0xd800 && cp <= 0xdbff) {
      const esc = this.s.slice(this.pos, this.pos + 2);
      const hex2 = this.s.slice(this.pos + 2, this.pos + 6);
      if (esc !== "\\u" || !/^[0-9a-fA-F]{4}$/.test(hex2)) this.fail();
      const lo = parseInt(hex2, 16);
      if (lo < 0xdc00 || lo > 0xdfff) this.fail();
      this.pos += 6;
      cp = 0x10000 + ((cp - 0xd800) << 10) + (lo - 0xdc00);
      return String.fromCodePoint(cp);
    }
    if (cp >= 0xdc00 && cp <= 0xdfff) this.fail(); // lone low surrogate
    return String.fromCharCode(cp);
  }

  obj(depth: number): Json {
    this.pos++;
    const out: { [k: string]: Json } = {};
    this.ws();
    if (this.s[this.pos] === "}") {
      this.pos++;
      return out;
    }
    for (;;) {
      this.ws();
      const k = this.string();
      this.ws();
      if (this.s[this.pos] !== ":") this.fail();
      this.pos++;
      this.ws();
      const v = this.value(depth + 1);
      if (k in out) throw new LexwattError("INVALID_INPUT"); // duplicate key
      out[k] = v;
      this.ws();
      const c = this.s[this.pos];
      this.pos++;
      if (c === "}") return out;
      if (c !== ",") this.fail();
    }
  }

  arr(depth: number): Json {
    this.pos++;
    const out: Json[] = [];
    this.ws();
    if (this.s[this.pos] === "]") {
      this.pos++;
      return out;
    }
    for (;;) {
      this.ws();
      out.push(this.value(depth + 1));
      this.ws();
      const c = this.s[this.pos];
      this.pos++;
      if (c === "]") return out;
      if (c !== ",") this.fail();
    }
  }
}

export function loads(text: string): Json {
  // reject lone surrogates (unpaired) — TextEncoder would silently fix them
  for (let i = 0; i < text.length; i++) {
    const cu = text.charCodeAt(i);
    if (cu >= 0xd800 && cu <= 0xdbff) {
      const lo = text.charCodeAt(i + 1);
      if (lo < 0xdc00 || lo > 0xdfff || Number.isNaN(lo)) throw new LexwattError("INVALID_INPUT");
      i++;
    } else if (cu >= 0xdc00 && cu <= 0xdfff) {
      throw new LexwattError("INVALID_INPUT");
    }
  }
  if (text.startsWith("﻿")) throw new LexwattError("INVALID_INPUT"); // BOM
  return new Parser(text).parse();
}

export function loadsBytes(data: Uint8Array): Json {
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(data);
  } catch {
    throw new LexwattError("INVALID_INPUT");
  }
  return loads(text);
}
